journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+          objId <- Sync[F].delay(UUID.randomUUID())
+          state1 = state.copy(
+            objects = state.objects + (objId -> AllocatedObject(obj, release, AllocatedObject.State.Used))
+          )
+          _ <- stateRef.set(state1)
+        } yield (objId, obj)
+      }
+
+      def checkForAcquireTimeout(
+        waiterId: WaiterId,
+        promise: Waiter[F, O]
+      ): F[Unit] = {
+        promise.tryGet.flatMap {
+          case Some(_) =>
+            Async[F].unit
+          case None =>
+            mutex.lock.surround {
+              promise.tryGet.flatMap {
+                case Some(_) =>
+                  Async[F].unit
+                case None =>
+                  for {
+                    _ <- promise.complete(
+                      Left(
+                        new TimeoutException(s"Resource acquisition timed out")
+                      )
+                    )
+                    _ <- stateRef.update { state =>
+                      state.copy(waiters = state.waiters - waiterId)
+                    }
+                  } yield ()
+              }
+            }
+        }
+      }
+
+      def registerWaiter(state: State[F, O], promise: Waiter[F, O]): F[Unit] = {
+        for {
+          waiterId <- Sync[F].delay(UUID.randomUUID())
+          _ <- stateRef.set {
+            state.copy(
+              waiters = state.waiters + (waiterId -> promise)
+            )
+          }
+
+          _ <- (Async[F].sleep(acquireTimeout) *> checkForAcquireTimeout(waiterId, promise)).start
+        } yield ()
+      }
+
+      val acquirePromise: F[Waiter[F, O]] =
+        Deferred[F, Either[Throwable, (ObjectId, O)]].flatMap { promise =>
+          mutex.lock.surround {
+            stateRef.get.flatMap { state =>
+              findFreeObject(state) match {
+                case Some((id, obj)) =>
+                  for {
+                    _ <- markObjectAsUsed(id, obj)
+                    _ <- promise.complete((id, obj.value).asRight)
+                  } yield ()
+
+                case None =>
+                  if (state.objects.size < poolSize) {
+                    for {
+                      o <- allocateObject(state)
+                      _ <- promise.complete(o.asRight)
+                    } yield ()
+                  } else {
+                    registerWaiter(state, promise)
+                  }
+              }
+            }
+          }.as(promise)
+        }
+
+      val acquireObject: F[(ObjectId, O)] = {
+        acquirePromise.flatMap { promise =>
+          promise.get.flatMap {
+            case Right(x) => x.pure[F]
+            case Left(err) => Async[F].raiseError(err)
+          }
+        }
+      }
+
+      def checkIfExpired(id: ObjectId, timeoutScheduledAt: FiniteDuration): F[Unit] = {
+        mutex.lock.surround {
+          stateRef.get.map(_.objects.get(id)).flatMap {
+            case Some(AllocatedObject(_, deallocate, AllocatedObject.State.Free(releasedAt))) =>
```

Comment:
As `checkIfEpired` is run async, then it is completely possible to arrive here while state is `AllocatedObject.State.Used` causing the incorrect error to be reported.

Reply:
fixed